lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+            });
+
+            Ok(spawner.spawn_local(processor_fut)?)
+        }
+
+        /// Returns a module controller for given module location.
+        ///
+        /// Reuses existing controller if possible.
+        /// Creates a new controller if needed.
+        pub async fn open_module(&self, loc:&module::Location) -> FallibleResult<module::StrongHandle> {
+            if let Some(existing_controller) = self.lookup_module(loc.clone()).await {
+                Ok(existing_controller)
+            } else {
+                Ok(self.create_module_controller(&loc).await?)
+            }
+        }
+
+        /// Retuns a text controller for given file path. It may designate
+        /// either the Luna source file or other file belonging to the project.
+        ///
+        /// File should be an existing, correct UTF-8 encoded file.
+        pub async fn open_file_text(&self, _path:&std::path::Path) -> FallibleResult<text::StrongHandle> {
+            // TODO [mwu] similar to the above (will need to add a second map
+            //      to keep handles or perhaps abstract the handles map to a
+            //      reusable structure?
+            //      Also might need to prepare a module first.
+            todo!()
+        }
+    }
+
+    /// Boilerplate wrappers over stateful `Data` APIs. Refer to the `Data`
+    /// methods for documentation.
+    impl StrongHandle {
+        #[allow(missing_docs)]
+        pub async fn lookup_module
+        (&self, loc:module::Location)
+         -> Option<module::StrongHandle> {
+            self.with(move |data| data.lookup_module(&loc)).await
+        }
+        #[allow(missing_docs)]
+        pub async fn insert_module
+        (&self, module:module::Data)
+         -> module::StrongHandle {
+            self.with(|data| data.insert_module(module)).await
+        }
+        #[allow(missing_docs)]
+        pub async fn call_fm<R>
+        (&self, f:impl FnOnce(&mut file_manager_client::Client) -> R)
+         -> R {
+            self.with(|data| f(&mut data.file_manager)).await
+        }
+    }
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+/// Module controller.
+pub mod module {
+    use super::*;
+
+    /// Structure uniquely identifying module location in the project.
+    /// Mappable to filesystem path.
+    #[derive(Clone,Debug,Eq,Hash,PartialEq)]
+    pub struct Location(pub String);
+    impl Location {
+        /// Obtains path (within a project context) to the file with this module.
+        pub fn to_path(&self) -> file_manager_client::Path {
+            // TODO [mwu] Extremely provisional. When multiple files support is
+            //            added, needs to be fixed, if not earlier.
+            let result = format!("./{}.luna", self.0);
+            file_manager_client::Path::new(result)
+        }
+    }
+
+    /// State data of the module controller.
+    #[derive(Clone,Debug)]
+    pub struct Data {
```

Comment:
The name "Data" is like "Utils" . It sohuld not be used. I know we are using it in other place, but there we use it as abstraciton for "any data which can be stored in a container". But here, the nam is really confusing.

Reply:
`wip/ao/text-controller`